contracts/src/lib/LibGateway.sol

```diff
@@ -416,6 +416,100 @@ library LibGateway {
         sendReceipt(crossMsg, success, ret);
     }
 
+    /// @notice executes a cross message if its destination is the current network, otherwise adds it to the postbox to be propagated further
+    /// This function assumes that the relevant funds have been already minted or burnt
+    /// when the top-down or bottom-up messages have been queued for execution.
+    /// This function is not expected to revert. If a controlled failure happens, a new
+    /// cross-message receipt is propagated for execution to inform the sending contract.
+    /// `Call` cross-messages also trigger receipts if they are successful.
+    /// @dev this method differs from `applyMsg` as it returns the return data to the caller
+    ///      instead of saving it. The vm will insert the ret back into the contract.
+    ///      At the current stage, we only call this method in top-down
+    ///      message execution. In bottom-up, the execution of cross net messages happens in 
+    ///      the parent, which most likely, we have no access to its vm execution 
```

Comment:
Why can't we handle it the same way in both cases? How does it work with Lotus?

Reply:
There are cases of "vm" errors which are not thrown in the contract, so receipt will not be available in the contract and cannot be committed.

Reply:
I have some recollection that we discussed once that it was tricky to reason about how to handle VM errors when a topdown "checkpoint" is handled as 3 independent transactions:
* the commitment of the parent finality
* the storing of the validator changes
* the execution of top-down messages

That is, it's not at all clear what it should mean if the first two succeed but the third fails.

As I understand this PR wants to skip failing messages one by one, even if they fail for unknown reasons. Another approach which IIRC I recommended was to have a single method to do everything, and fail or succeed atomically; if any of the three aspects fail, the entire proposed top-down execution fails, and we can try again later once we fixed the bug or something. 

It ain't pretty, but better than a partial execution which cannot be retried.